src/common/mfu_flist_copy.c

```diff
@@ -379,7 +396,7 @@ static int mfu_copy_ownership(
     gid_t gid = (gid_t) mfu_flist_file_get_gid(flist, idx);
 
     /* note that we use lchown to change ownership of link itself, it path happens to be a link */
-    if(mfu_lchown(dest_path, uid, gid) != 0) {
+    if(mfu_file_lchown(dest_path, uid, gid, mfu_dst_file) != 0) {
```

Comment:
Do we also need a dereference/no-dereference option for chown?

Reply:
I found that we don't have to worry about dereference/no-dereference for the _**destination**_, since the destination is created as a mirror of the source, and when we stat/lstat the source, we save the type (file/dir/link).
For example, if we use `--dereference` on a source symbolic link, `stat` will be called, which will dereference the link and store `struct stat` info for the file/dir (not the link). Then, we create the destination based on this `struct stat` info. Then, when we call `lchown` on the _**destination**_, it is okay to not dereference the link again, since if we originally dereferenced the source link then the destination wouldn't even be a link at all. And if the destination is a link, then we must not have dereferenced links, so `lchown` again is correct.

Reply:
But then again... please let me know if there are holes in this that I'm not seeing.

Reply:
Your comment got me thinking... And I realized that I didn't account for `dcp --input` (I haven't used that option at all). This function will need to handle dereferencing:
https://github.com/hpc/mpifileutils/blob/4ec784108d066abe5060ebb197c1dba83e88bd7d/src/common/mfu_flist_walk.c#L763-L769

Reply:
> 
> 
> I found that we don't have to worry about dereference/no-dereference for the _**destination**_, since the destination is created as a mirror of the source, and when we stat/lstat the source, we save the type (file/dir/link).
> For example, if we use `--dereference` on a source symbolic link, `stat` will be called, which will dereference the link and store `struct stat` info for the file/dir (not the link). Then, we create the destination based on this `struct stat` info. Then, when we call `lchown` on the _**destination**_, it is okay to not dereference the link again, since if we originally dereferenced the source link then the destination wouldn't even be a link at all. And if the destination is a link, then we must not have dereferenced links, so `lchown` again is correct.

Yes, thanks for the explanation.  I also just realized that under closer inspection.  You're right in that the destination should be the right type already, so it's fine to always use the "l" variant of the calls for destination items.  It only makes a difference on source items.

Reply:
Ah yes, good catch on ``mfu_flist_stat``.

Not all tools calls ``flist_stat`` after reading an input file.  If someone reads an input file as we do not call ``flist_stat``, then we may end up with a source list that does still have symlinks, e.g., if they ran dwalk without dereference and saved that to a file with ``--output`` which they then read back in to another tool.  We'll want to work through that case, though nothing jumps to mind right away.

Reply:
I can't get `--input` to work, due to some other issues.
- https://github.com/daltonbohning/mpifileutils/blob/c409987e1a572bc3e64337add489980942f8fe58/src/dcp/dcp.c#L933-L941
  - I introduced this, without realizing that `--input` allows input from a file. This is easy to fix, and I will
    - `if ((argc-optind) < 2 && !inputname)`

- https://github.com/daltonbohning/mpifileutils/blob/c409987e1a572bc3e64337add489980942f8fe58/src/dcp/dcp.c#L1000-L1011
  - This essentially requires both a `src` and `dst` to be provided, even when using `--input`. I don't believe I originally introduced this, so it seems it must have already been non-functioning. (not sure about pre-DAOS)
  - This is easy to fix, and I will: `if (!valid and !inputname)`

- https://github.com/daltonbohning/mpifileutils/blob/c409987e1a572bc3e64337add489980942f8fe58/src/common/mfu_flist_walk.c#L766
  - We are trying to reference `CURRENT_PFILE`, but that might global not always be initialized.
  - I'll fix this as well.

In short... I think some of our DAOS changes broke the `--inputname` functionality, so I will correct that.
And then I will get `--inputname` to work correctly with `--dereference`

Reply:
Okay, I corrected that issue and also adjusted `mfu_flist_stat` to accept `dereference` parameter, and also try to automatically dereference symbolic links that were dereferenced in `dwalk`'s output. The idea is that if we come across something that _**is not**_ a link, then we go ahead and try to dereference that, since `dwalk` might output `S_ISREG` for a symlink path, which means that the symlink should actually be dereferenced.

I will need to look at other places the `dwalk`'s output is used as input, and make sure these places also correctly handle when the path was dereferenced by `dwalk`.

Reply:
These are places that might need consideration to handle `dwalk --dereference` input.
- https://github.com/daltonbohning/mpifileutils/blob/4f53ca409712a2ba898dd14cd9f97d6c46dbcdde/src/common/mfu_flist_chmod.c#L1010
  - I think `chmod` doesn't matter, since `chmod` on a link dereferences the link anyway.
- https://github.com/daltonbohning/mpifileutils/blob/copy-dereference/src/common/mfu_flist_remove.c
  - Maybe somewhere in here, but this is more complex
  - `dsh`, `dsync`, and `drm` call `mfu_flist_unlink`
  - Maybe we just don't want to allow deleting in this way, since it could be dangerous to dereference and delete something accidentally.
- https://github.com/daltonbohning/mpifileutils/blob/copy-dereference/src/dreln/dreln.c
  - I think `dreln` is actually okay, since it manually filters out only links. If you pass it an input that has dereferenced links, then it just won't do anything at all.
- https://github.com/daltonbohning/mpifileutils/blob/4f53ca409712a2ba898dd14cd9f97d6c46dbcdde/src/common/mfu_param_path.c#L632
  - `mfu_param_path_set` / `mfu_param_path_set_all`
  - `dfind`, among others, call this to get `stat` info, among other things. This one seems a little complex to get working entirely, since the `stat` info is not always used entirely. For example, even `dcp` calls `mfu_param_path_set_all` on the source paths, but then `mfu_param_path_check_copy` still behaves as expected.
  - I'm not sure of the other implications of this for `dfind`

If this is too much to get into at this time, we could either leave it as is, which means `--input` might not always be dereferenced (at least not until those tools get explicit `--dereference` support), or I could back-out `--dereference` from `dwalk` until we are better prepared. I've really only working with `dcp`, so I'm not knowledgeable about the other tools.